### DISCUSSION TITLE
Can JsValue memory leak?

### DISCUSSION THREAD

**Discussion by oblique** [wasm-bindgen/wasm-bindgen#4017] [Q&A] [answered]
I'm working a a project which moves a lot of data to JS. The data are initially Rust structs and then converted to `JsValue`  by using `serde-wasm-bindgen` and send to JS, however I notice that memory usage grows a lot over time.

My questions:

1. If `JsValue` is sent to the JS. Will garbage collector deallocate it?
2. If `JsValue` is not sent to the JS, but dropped in Rust. Will it memory leak or deallocated by GC?
3. If `JsValue` is sent from JS to Rust and is dropped. Will garbage collector deallocate it?
4. If 2 or 3 memory leak, how can I free the memory?

**Comment by daxpedda** [comment, marked as answer]
Generally speaking, all the cases you mentioned, should be handled by the GC.
However there are some subtle issues that can't really be resolved because of the limitations surrounding JS, see #3917.

All-in-all, unless something is intentionally not cleaning up memory, everything should be handled correctly by the GC *eventually*.
If not, you have likely found a bug. Let us know if you do!

(I'm not familiar with `serde-wasm-bindgen`)

**Reply by daxpedda** [reply]
Forgot to mention #3854 as well, so if you encountered that it will be fixed in the next release.

**Reply by caiges** [reply]
Out of curiosity, how is the decision to release made?

**Reply by daxpedda** [reply]
There is no specific process in place, its done on request.

It has already been requested, including by myself, so I'm currently working on the last couple of things before the release.
I'm hoping today? But seems I've said that two days ago already, so wish me luck!

**Reply by oblique** [reply]
Just in the case that any other poor soul is looking for the memory leak but can not find it: The memory leak for me was because of [tracing-web's performance layer](https://docs.rs/tracing-web/0.1.3/tracing_web/struct.PerformanceEventsLayer.html). It was calling `Performance.mark()` and `Performance.measure()` which creates JS objects that live throughout the whole process. This is not a bug in `tracing-web` but it is how performance marks work.
